integr8ly/tutorial-web-app-operator
# INTLY-3245 - Logic to upgrade webapp image on operator upgrade

## Description
Upgrades the webapp image on webapp operator upgrade if there is a difference between the current deployed image and the image as specified in the operator

Associated pull requests;
* https://github.com/integr8ly/integreatly-operator/pull/170
* https://github.com/integr8ly/manifests/pull/85

## Verification
### The short version:
This should be sufficient to test that the web app image pod/deployment config would be reconciled/upgraded if there is an operator upgrade that points to a different web app image version

* Create OperatorSource pointed to my registry
```
apiVersion: operators.coreos.com/v1
kind: OperatorSource
metadata:
  name: integreatly-operators
  namespace: openshift-marketplace
spec:
  authorizationToken: {}
  displayName: Integreatly Operators
  endpoint: 'https://quay.io/cnr'
  publisher: Integreatly Publisher
  registryNamespace: kevfan
  type: appregistry
```
* Create test namespace
  * e.g. `test-webapp-ns`
* Install `Tutorial Web App Operator` into test namespace `test-webapp-ns`
* Create Web App CR:
```
apiVersion: "integreatly.org/v1alpha1"
kind: "WebApp"
metadata:
  name: "tutorial-web-app-operator"
  labels:
    app: "tutorial-web-app"
spec:
  app_label: "tutorial-web-app"
  template:
    path: "/home/tutorial-web-app-operator/deploy/template/tutorial-web-app.yml"
    parameters:
      OPENSHIFT_OAUTHCLIENT_ID: "tutorial-web-app"
      OPENSHIFT_HOST: ""
      SSO_ROUTE: ""
```
* After pods are provisioned, edit `tutorial-web-app` deployment config container image to:
  * `quay.io/integreatly/tutorial-web-app:latest`
* Verify that the image is reconciled to `quay.io/integreatly/tutorial-web-app:2.19.1`

### The Long Version
This would verify on a integreatly install a tutorial web app operator upgrade would upgrade the web app image and that the upgrade doesn't affect any saved progress in the walkthroughs

* Push webapp operator to your own registry from of manifest repo (should push `v0.0.32` of operator as the current csv)
  * Integreatly operator would also need to be deployed also due to https://github.com/integr8ly/integreatly-operator/pull/170/files 
    * Would involve building integreatly-operator image, pushing to your own registry, updating integreatly manifest to point to image and push the integreatly csv to your own registry
* Install integreatly onto cluster using your own registry
* Go through walkthrough 1
* Build new web app operator version pointing to a different image and push to own registry:
  * e.g. `quay.io/integreatly/tutorial-web-app:latest`
* Build new csv web app version (`0.0.33`) using newly built image in the previous step and push to own registry 
* Upgrade the operator
  * Can be caused by removing `status` field in operator source 
  * Also may need to be approved in operator management
* Check that the web-app image pod gets upgraded to the new image
  * Can be checked in the deployment config / pod yaml
* Verify walkthrough 1 progress is still the same as previous and the integration works as previous
